server/src/main/java/org/elasticsearch/action/admin/cluster/coordination/TransportClusterFormationInfoAction.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.admin.cluster.coordination;
+
+import org.elasticsearch.action.FailedNodeException;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.nodes.TransportNodesAction;
+import org.elasticsearch.cluster.coordination.Coordinator;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.tasks.CancellableTask;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Transport action for retrieving information regarding the cluster formation.
+ * This will fetch the information from the requested nodes, and it'll likely be sourced from
+ * {@link Coordinator}.
+ */
+public class TransportClusterFormationInfoAction extends TransportNodesAction<
```

Comment:
I just realised that this probably can't be a `TransportNodesAction`: we're using this action when not a proper cluster so there's nothing holding the connections to other nodes open, so we'll have to explicitly acquire/release the connection yourself. See e.g. how `Coordinator#handleJoinRequest` and `HotThreadsLoggingLagListener#onLagDetected` wrap their transport activities in a call to `TransportService#connect` and release the connection afterwards. `TransportNodesAction` won't do that, we probably need to call `TransportService#sendRequest` directly.

Also it won't work to use the current cluster state to resolve node IDs since there is no cluster when using this. We'll need to use the exact `DiscoveryNode` instances that discovery tells us about.